File: src/hyb-simple/hyb_object.cc
#include "hyb-simple/hyb_object.hh"

#include <iostream>

HybObject::HybObject(HybObjectParams *params) : SimObject(params), event([this]{processEvent();}, name()),latency(100),timesLeft(10)
{
    std::cout << "[HybObject 构造方法]" << std::endl;
}

void
HybObject::processEvent()
{
    timesLeft--;
    DPRINTFNR("[processEvent] Processing the event! %d left\n", timesLeft);

    if (timesLeft <= 0) {
        std::cout << "[processEvent] Done firing!" << std::endl;
    } else {
        schedule(event, curTick() + latency);
    }
}

void
HybObject::startup()
{
    schedule(event, latency);
}

HybObject*
HybObjectParams::create()
{
    return new HybObject(this);
}

File: src/hyb-simple/HybObject.py
from m5.params import *
from m5.SimObject import SimObject

class HybObject(SimObject):
    type = 'HybObject'
    cxx_header = "hyb-simple/hyb_object.hh"
    time_to_wait = Param.Latency("Time before firing the event")
    number_of_fires = Param.Int(10, "Number of times to fire the event before "
                                   "goodbye")